Unvendor python runtime from Greenplum installation directory

tylarb opened the pull request:
Currently, Greenplum is built with a vendored python in $GPHOME/lib/python, which is then set to be the PYTHONHOME environment value after sourcing `greenplum_path.sh` 

This PR removes the requirement to vendor python in this directory, and by extension, unsets the PYTHONHOME environment variable.

This has several implications to the build and installation process.

With regards to building:
- previously, our vendored python provided the python headers necessary to compile pl/python. Now, these headers must be provided by python-devel (in RHEL or Suse environments) or python-dev (in Debian-based systems) to compile Greenplum.
- The python libraries that we ship, which include pygresql and psutil, should be compiled with system gcc - the same gcc version which compiled python that they are running against. Therefore, while a developer may be compiling greenplum with a higher gcc version, a system gcc needs to be installed in order to build the vendored python libraries and complete building Greenplum.
> PL/python is still built with the compiler which builds Greenplum.

With regards to running:
- A system python 2.7 is now required to run greenplum. This is not an issue on any current officially supported platform (Ubuntu 18.04, SLES 12, or RHEL/Centos 7) as they ship with a compatible python. However, some systems like Ubuntu 20.04 will need to install a legacy python 2 for the time being in order to run the utilities.

The scope of this PR is two commits:
1. Unvendoring python, which is largely done in the Makefiles, especially gpAux Makefile, and some related unsetting of PYTHONHOME and adjusting the PYTHONPATH. 
2. CI improvements to make both Unvendoring python possible and toolchain improvements to make it easier, which includes using a "system provided" gcc and g++ instead of one built from source. The PR in gp-image-baking to make these changes is here: https://github.com/pivotal/gp-image-baking/pull/76

Once the gp-image-backing PR is merged, the "use dev images" commit will be dropped, using "latest" images in gpdb7 master.

The completed .rpm or .deb in the build process will also need to specify a python version as a required dependency.

tylarb commented:
a dev pipeline showing this work is available here: 
https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-unvendor_python

tylarb commented:
We need to update the PR pipeline as well with these changes.

d commented on concourse/pipelines/templates/gpdb-tpl.yml:
```diff
@@ -531,71 +531,44 @@ resources:
   source:
     repository: pivotaldata/gpdb7-centos7-build
-    tag: latest
+    tag: devv13
```
is this change in image tags intentional? are we committing this?

tylarb replied:
"Once the gp-image-backing PR is merged, the "use dev images" commit will be dropped, using "latest" images in gpdb7 master."  - image baking PR here: https://github.com/pivotal/gp-image-baking/pull/76

bhuvnesh2703 commented on gpMgmt/bin/generate-greenplum-path.sh:
```diff
@@ -40,31 +40,16 @@ if [ -h \${GPHOME}/../greenplum-db ]; then
-else
-    PYTHONPATH="\$GPHOME/lib/python:${PYTHONPATH}"
-fi
+PYTHONPATH="\$GPHOME/lib/python:${PYTHONPATH}"
```
is this required? or system path is good enough

tylarb replied:
For the python libraries we vendor, yes. This include pygresql and psutil. Additional libraries can be loaded from the OS path, hence $GPHOME libraries first, followed by any additional ones

bhuvnesh2703 commented:
LGTM. just a minor comment..

d commented on gpMgmt/bin/Makefile:
```diff
@@ -80,13 +80,13 @@ pygresql:
 	    cd $(PYLIB_SRC)/$(PYGRESQL_DIR) && DESTDIR="$(DESTDIR)" CC=cc python setup.py build; \
 	else \
-	    cd $(PYLIB_SRC)/$(PYGRESQL_DIR) && DESTDIR="$(DESTDIR)" CC="$(CC)" python setup.py build; \
+	    cd $(PYLIB_SRC)/$(PYGRESQL_DIR) && DESTDIR="$(DESTDIR)" env -u CC python setup.py build; \
```
This feels pretty defensive. A comment about _why_ this is necessary will help us 2 weeks from now...

For context, I've been running a different compiler (Clang 11) from the system compiler that's used to build Python and I haven't run into any problem yet...

tylarb replied:
The comment on this should be in the commit message - I verified. Here's the copy:

```   
     - Finally, certain tools may specify a compiler version - for example,
      python packages might look for a system provided gcc rather than a
      vendored one. Ensuring that "gcc" is the correct version for the rest of
      GPDB compilation and in the path can break certain installations once
      they are unvendored from gpAux.
    
    - However, python packages should be compiled with the same GCC version
      that compiled system python - thus, we unset the "CC" variable when
      installing additional python libraries.
```

I will amend the commit with this detail:
Specifically, the compiler flags used to compile _python_ are saved and used to compile _libraries_ when the `python setup.py build` process is used. So if system python has different compiler flags than might be allowed in a newer version of gcc, the build of the libraries fails. 

This has no impact on any Postgres code, nor on PL/Python. Only the python libraries which we vendor, which are now running under system python. These are also targeted to be removed from vendoring in a future commit (i.e. install via pip or otherwise)

d commented:
> * The python libraries that we ship, which include pygresql and psutil, should be compiled with system gcc - the same gcc version which compiled python that they are running against. Therefore, while a developer may be compiling greenplum with a higher gcc version, a system gcc needs to be installed in order to build the vendored python libraries and complete building Greenplum.

Why? A comment explaining what would fail if we don't require _both_ compilers to be present will help us two weeks from now (or more likely, 18 months from now)

d commented on concourse/scripts/unit_tests_gpcloud.bash:
```diff
@@ -8,7 +8,6 @@ function gen_env(){
 	cat > ~/run_unit_tests.sh <<-EOF
 	set -exo pipefail
 
-	source /opt/gcc_env.sh
```
good riddance

d commented on concourse/scripts/run_behave_test.sh:
```diff
@@ -6,13 +6,16 @@ source "${CWDIR}/common.bash"
   alias mdd='cd \$MASTER_DATA_DIRECTORY'
+  # pip installs are done using --user, so are in ~/.local/bin which
+  # is not in the default path over ssh
+  export PATH=~/.local/bin:${PATH}
```
ewww

d commented on concourse/scripts/ic_gpdb.bash:
```diff
@@ -29,9 +29,6 @@ function gen_env(){
 		source /usr/local/greenplum-db-devel/greenplum_path.sh
-		if [ -f /opt/gcc_env.sh ]; then
-		    source /opt/gcc_env.sh
-		fi
```
:clap:

d commented on concourse/scripts/common.bash:
```diff
@@ -56,84 +52,28 @@ function make_cluster() {
-    deactivate
+
+    pip install --user -r ${requirements_txt}
+    while read -r host; do
```
Out-of-scope question: while not the fault of this patch, this `pip install`-in-a-loop gives me pause:

1. It just does a `pip install --user` which mixes the set of library versions with what a user may already have. Have considered solutions in the future that would mitigate the risk of mixing with users' libraries?
1. In order to give us a consistent set of dependencies, this also assumes (which as of writing is not false) that the `requirements.txt` uses an exhaustive and strictly pinning list of `librar==x.y` (otherwise different host may perform different requirement resolution, given the difference in time), that makes the developers' lives much harder when they want to bump any particular dependency in the future. Have we consider something like `Pipfile` and `Pipefile.lock` in the future?

tylarb replied:
> It just does a pip install --user which mixes the set of library versions with what a user may already have. 

The pip install here is ONLY for testing the code, inside of concourse, for behave tests specifically. While I understand the concern of mixing with user libraries, behave tests in particular require a running instance of Greenplum. Apart from running the tests in a container, one might also use a virtual env to install the specific packages to avoid conflict with the user's installed packages.

In this case, a virtual env is not necessary, as we can control the testing infrastructure in concourse. The reason a virtual env was used before is because the installed packages needed to be linked against (and compiled with the same compiler as) the vendored version of python - hence the patchelf requirement as well.

I'll add a comment to the OSS readme/test docs that a virtual env is required in any case. I expected some documentation fixups to be necessary. 

> That makes the developers' lives much harder when they want to bump any particular dependency in the future. Have we consider something like Pipfile and Pipefile.lock in the future?

I agree. Clearly, none of this code has been updated in any recent time, so the pain point has also not been addressed.

I don't know if we will be required to address this in the scope of our work, but rest assured that, when we hit this pain point, it will be addressed as part of the tech-debt scope. Similarly to how we have addressed GCC compiler vendoring in this PR.

tylarb commented:
@d I've updated commit message to be more clear, and rebased off of master. also updated developer docs in gpMgmt/bin/README, now renamed gpMgmt/bin/README.md

d commented:
Was there any reason we did _not_ drop the last commit "Use dev images for pipeline" when we finalize this PR? @tylarb

tylarb commented:
@d I made a mistake when I merged. I added a fix here https://github.com/greenplum-db/gpdb/commit/19200adc22a2cb973818dde21cb974f24e4a07f7

d commented:
Awesome thanks!
